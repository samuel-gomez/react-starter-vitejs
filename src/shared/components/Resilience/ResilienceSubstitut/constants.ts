export const RESILIENCE_MODE = {
  alert: 'alert',
  fallback: 'fallback',
  none: 'none',
} as const;

export const DEFAULT_CLASS_ALERT = 'af-alert';
export const DEFAULT_CLASS_CONTAINER = 'container';
export const WITH_ACTION_CLASS_MODIFIER = 'with-action';
