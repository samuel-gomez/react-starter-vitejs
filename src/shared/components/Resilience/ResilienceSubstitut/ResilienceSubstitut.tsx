import { ElementType, ReactNode } from 'react';
import Alert from '@axa-fr/react-toolkit-alert/dist/esm/index';
import Button from '@axa-fr/react-toolkit-button/dist/esm/index';
import type { Tanomaly } from 'shared/types';
import { ERESILIENCE_MODE, DEFAULT_CLASS_ALERT, DEFAULT_CLASS_CONTAINER } from './constants';

export const Empty = () => null;

type TsetClassModifier = {
  type: string;
  resilienceModifier: string;
};

export const setClassModifier = ({ type, resilienceModifier }: TsetClassModifier) =>
  [type, resilienceModifier !== '' ? resilienceModifier : ''].join(' ').trim();

type TsetClassName = {
  classAlertCt?: string;
  classContainerCt?: string;
  resilienceModifier: string;
};

export const setClassName = ({ resilienceModifier, classAlertCt = DEFAULT_CLASS_ALERT, classContainerCt = DEFAULT_CLASS_CONTAINER }: TsetClassName) =>
  resilienceModifier === classContainerCt ? `${classContainerCt} ${classAlertCt}` : classAlertCt;

type TResilienceSubstitut<Trefetch> = {
  anomaly: Tanomaly;
  refetch?: Trefetch;
  children?: ReactNode;
  resilienceMode?: keyof typeof ERESILIENCE_MODE;
  FallbackComponent?: ElementType;
  resilienceModifier?: string;
  setClassModifierFn?: typeof setClassModifier;
  setClassNameFn?: typeof setClassName;
};

const ResilienceSubstitut = <Trefetch extends React.MouseEventHandler<HTMLButtonElement>>({
  anomaly,
  refetch,
  children,
  resilienceMode = ERESILIENCE_MODE.alert,
  FallbackComponent = Empty,
  resilienceModifier = '',
  setClassModifierFn = setClassModifier,
  setClassNameFn = setClassName,
}: TResilienceSubstitut<Trefetch>) => {
  const { label, detail = '', type = 'error', iconName = 'exclamation-sign' } = anomaly;
  const classModifier = setClassModifierFn?.({ type, resilienceModifier });
  const className = setClassNameFn?.({ resilienceModifier });

  return {
    [ERESILIENCE_MODE.alert]: (
      <Alert className={className} title={label} icon={iconName} classModifier={classModifier}>
        {detail && <p>{detail}</p>}
        {refetch && (
          <Button aria-label="Réessayer" type="button" className="af-link" onClick={refetch}>
            <span className="af-link__text">Réessayer</span>
          </Button>
        )}
        {children}
      </Alert>
    ),
    [ERESILIENCE_MODE.none]: null,
    [ERESILIENCE_MODE.fallback]: <FallbackComponent />,
  }[resilienceMode];
};

export default ResilienceSubstitut;
