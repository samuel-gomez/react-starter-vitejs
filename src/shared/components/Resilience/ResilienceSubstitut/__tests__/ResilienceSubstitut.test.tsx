import { render, screen } from '@testing-library/react';
import ResilienceSubstitut, { setClassModifier, setClassName } from '../ResilienceSubstitut';

describe('<ResilienceSubstitut/>', () => {
  const defaultProps = {
    anomaly: {
      label: 'label',
      detail: 'detail',
    },
  };

  it('Render <ResilienceSubstitut/> with default props and return Alert resilience', () => {
    const { container } = render(<ResilienceSubstitut {...defaultProps} />);
    expect(screen.getByText('label')).toBeDefined();
    expect(screen.getByText('detail')).toBeDefined();
    expect(container.querySelector('.glyphicon-exclamation-sign')).toBeDefined();
  });

  it('Render <ResilienceSubstitut/> with default props and refetchMock', () => {
    const refetchMock = vi.fn();
    const { container } = render(<ResilienceSubstitut {...defaultProps} refetch={refetchMock} />);
    expect(screen.getByText('RÃ©essayer')).toBeDefined();
    expect(container.querySelector('.af-alert--with-action')).toBeDefined();
    expect(container.querySelector('.af-restitution--with-action')).toBeDefined();
    expect(container.querySelector('.af-btn--hasiconRight')).toBeDefined();
  });

  it('Render <ResilienceSubstitut/> with classModifier="simple" and return Alert resilience with simple classModifier', () => {
    const customProps = { ...defaultProps, classModifier: 'simple', anomaly: { ...defaultProps.anomaly, iconName: 'icon2', type: 'warning' } };
    const { container } = render(<ResilienceSubstitut {...customProps} />);
    expect(screen.getByText('label')).toBeDefined();
    expect(screen.getByText('detail')).toBeDefined();
    expect(container.querySelector('.af-alert--simple')).toBeDefined();
    expect(container.querySelector('.glyphicon-icon2')).toBeDefined();
  });

  it('Render <ResilienceSubstitut/> with resilienceMode="none" and return {}', () => {
    const { container } = render(<ResilienceSubstitut {...defaultProps} resilienceMode="none" />);
    expect(container.firstChild).toBeNull();
  });

  it('Render <ResilienceSubstitut/> with resilienceMode="fallback" and FallBackComponent = Custom and Should contain Custom element', () => {
    const Custom = () => <div>FallbackComponent</div>;
    render(<ResilienceSubstitut {...defaultProps} resilienceMode="fallback" FallbackComponent={Custom} />);
    expect(screen.getByText('FallbackComponent')).toBeDefined();
  });
});

describe('setClassModifier', () => {
  it('Should return "error" when setClassModifier called type="error" and classModifier=""', () => {
    const result = setClassModifier({ type: 'error', classModifier: '' });
    const expected = 'error';
    expect(result).toEqual(expected);
  });
  it('Should return "error" when setClassModifier called type="error" and classModifier=undefined', () => {
    const result = setClassModifier({ type: 'error' });
    const expected = 'error';
    expect(result).toEqual(expected);
  });

  it('Should return "error mymodifier" when setClassModifier called type="error" and classModifier="mymodifier"', () => {
    const result = setClassModifier({ type: 'error', classModifier: 'mymodifier' });
    const expected = 'error mymodifier';
    expect(result).toEqual(expected);
  });
});

describe('setClassName', () => {
  it('Should return "af-alert" when setClassName called with newClassModifier=""', () => {
    const result = setClassName({ newClassModifier: '' });
    expect(result).toEqual('af-alert');
  });

  it('Should return "container af-alert" when setClassName called with newClassModifier="container"', () => {
    const result = setClassName({ newClassModifier: 'container' });
    const expected = 'container af-alert';
    expect(result).toEqual(expected);
  });
});
