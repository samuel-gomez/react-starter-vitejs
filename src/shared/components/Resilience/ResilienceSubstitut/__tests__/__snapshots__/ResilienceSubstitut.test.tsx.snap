// Vitest Snapshot v1

exports[`<ResilienceSubstitut/> > Render <ResilienceSubstitut/> with classModifier="simple" and return Alert resilience with simple classModifier 1`] = `
<DocumentFragment>
  <div
    class="af-alert af-alert--warning"
    role="alert"
  >
    <div
      class="af-alert__title"
    >
      <div
        class="af-alert__title-icon"
      >
        <i
          class="glyphicon glyphicon-icon2"
        />
      </div>
      <div
        class="af-alert__title-text"
      >
        label
      </div>
    </div>
    <div
      class="af-alert__content"
    >
      <div
        class="af-alert__content__left"
      />
      <div
        class="af-alert__content__right"
      >
        <p>
          detail
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ResilienceSubstitut/> > Render <ResilienceSubstitut/> with default props and refetchMock 1`] = `
<DocumentFragment>
  <div
    class="af-alert af-alert--error"
    role="alert"
  >
    <div
      class="af-alert__title"
    >
      <div
        class="af-alert__title-icon"
      >
        <i
          class="glyphicon glyphicon-exclamation-sign"
        />
      </div>
      <div
        class="af-alert__title-text"
      >
        label
      </div>
    </div>
    <div
      class="af-alert__content"
    >
      <div
        class="af-alert__content__left"
      />
      <div
        class="af-alert__content__right"
      >
        <p>
          detail
        </p>
        <button
          aria-label="Réessayer"
          class="af-link"
          type="button"
        >
          <span
            class="af-link__text"
          >
            Réessayer
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ResilienceSubstitut/> > Render <ResilienceSubstitut/> with default props and return Alert resilience 1`] = `
<DocumentFragment>
  <div
    class="af-alert af-alert--error"
    role="alert"
  >
    <div
      class="af-alert__title"
    >
      <div
        class="af-alert__title-icon"
      >
        <i
          class="glyphicon glyphicon-exclamation-sign"
        />
      </div>
      <div
        class="af-alert__title-text"
      >
        label
      </div>
    </div>
    <div
      class="af-alert__content"
    >
      <div
        class="af-alert__content__left"
      />
      <div
        class="af-alert__content__right"
      >
        <p>
          detail
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ResilienceSubstitut/> > Render <ResilienceSubstitut/> with resilienceMode="fallback" and FallBackComponent = Custom and Should contain Custom element 1`] = `
<DocumentFragment>
  <div>
    FallbackComponent
  </div>
</DocumentFragment>
`;
